### PR TITLE
TLS certificate and key rotation.

### PR DESCRIPTION
**Pull request by neunhoef**

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8261/

File: arangod/GeneralServer/GeneralServer.cpp [hunk 1]
```diff
@@ -185,12 +185,28 @@ IoContext& GeneralServer::selectIoContext() {
...
     _sslContext.reset(new asio_ns::ssl::context(SslServerFeature::SSL->createSslContext()));
   }
-  return *_sslContext;
+  return _sslContext;
+}
+
+Result GeneralServer::reloadTLS() {
+  std::lock_guard<std::mutex> guard(_sslContextMutex);
+  try {
+    _sslContext.reset(new asio_ns::ssl::context(SslServerFeature::SSL->createSslContext()));
```
Comment (graetzer) [inline arangod/GeneralServer/GeneralServer.cpp]: doesn't this copy the context at least once

Comment (neunhoef) [reply arangod/GeneralServer/GeneralServer.cpp]: Right. I changed the helper to return shared_ptrs right away.

File: arangod/GeneralServer/SslServerFeature.cpp [hunk 2]
```diff
@@ -592,3 +597,90 @@ std::string SslServerFeature::stringifySslOptions(uint64_t opts) const {
...
+    }
+  }
+}
+
+// Dump all SSL related data into a builder, private keys
+// are hashed.
+Result SslServerFeature::dumpTLSData(VPackBuilder& builder) const {
+  { VPackObjectBuilder guard(&builder);
+    dumpPEM(_keyfileContent, builder, "keyfile");
+    dumpPEM( _cafileContent, builder, "clientCA");
```
Comment (graetzer) [inline arangod/GeneralServer/SslServerFeature.cpp]:
```suggestion
    dumpPEM(_cafileContent, builder, "clientCA");
```

File: arangod/GeneralServer/SslServerFeature.cpp [hunk 3]
```diff
@@ -592,3 +597,90 @@ std::string SslServerFeature::stringifySslOptions(uint64_t opts) const {
   // strip initial comma
   return result.substr(2);
 }
+
+
+static void splitPEM(std::string const& pem, std::vector<std::string>& certs, std::vector<std::string>& keys) {
+  std::vector<std::string> result;
```
Comment (graetzer) [inline arangod/GeneralServer/SslServerFeature.cpp]: are PEM files always in this format ?

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8382/

File: arangod/GeneralServer/SslServerFeature.cpp (hunk 3 above)
Comment (neunhoef) [reply arangod/GeneralServer/SslServerFeature.cpp]: As far as I know, yes. There might be people who use some other file and call it .pem, but everything I have seen so far and the definition of PEM is this.

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8392/

**Comment by neunhoef** [comment]
Note that the corresponding docs are in this PR: https://github.com/arangodb/docs/pull/321

**Comment by neunhoef** [comment]
The red in the jenkins run is unrelated. This is good to be merged as soon as we have a review.

**Pull request merged**